File: library/src/main/java/renetik/android/lang/CSRunConsolidator.java
package renetik.android.lang;

import static renetik.android.lang.CSLang.doLater;
import static renetik.android.lang.CSLang.list;
import renetik.java.collections.CSList;
import renetik.java.callback.CSRun;

public class CSRunConsolidator implements CSRun {

	private final CSList<CSRun> runnables = list();
	private final int miliseconds;
	private boolean isRunning;

	public CSRunConsolidator(int miliseconds) {
		this.miliseconds = miliseconds;
	}

	public void invoke(CSRun runnable) {
		if (isRunning)
			runnables.add(runnable);
		else {
			runnable.run();
			isRunning = true;
			doLater(miliseconds, this);
		}
	}

	@Override
	public void run() {
		if (runnables.getHasItems()) {
			runnables.removeLast().run();
			doLater(miliseconds, this);
		} else isRunning = false;
	}

}

File: library/src/main/java/renetik/android/lang/CSRunConsolidator.kt
package renetik.android.lang

import renetik.java.collections.list
import renetik.java.lang.CSLang

class CSRunConsolidator(private val miliseconds: Int) {

    private val runnables = list<() -> Unit>()
    private var isRunning = CSLang.NO

    operator fun invoke(runnable: () -> Unit) {
        if (isRunning) runnables.add(runnable)
        else {
            runnable()
            isRunning = CSLang.YES
            doLater(miliseconds) { run() }
        }
    }

    fun run() {
        if (runnables.hasItems) {
            runnables.deleteLast()!!.invoke()
            doLater(miliseconds) { run() }
        } else isRunning = CSLang.NO
    }

}
